### DISCUSSION TITLE
Controlling external motors

### DISCUSSION THREAD

**Discussion by guykoy123** [pyControl/code#80] [Q&A] [answered]
I am creating an experiment setup where 3 stepper motors (by Zaber) connected to linear actuators (x,y and z axis) move a trigger object to the mouse.
The motors can be connected with a mini-DIN or with usb adaptor to a pc. They are controlled with a serial connection.
[Link to motor specs](https://www.zaber.com/manuals/T-LSR#m-7-11-home-cmd-1)
Is there a way to connect it to the pyControl board? or alternatively send a command from the board to computer to trigger the movement?
i thought of using micropython but i can't figure out how to run code on the micropython board.
I know i can define custom devices using the pycontrol library how ever i do not have the knowledge in designing a custom PCB for it (any leads to learn that may help as well)
Thanks

**Comment by ThomasAkam** [comment, marked as answer]
Hi @guykoy123,

Apologies for the delay getting to this.  

It should be possible to control these linear actuators from a pyControl setup.   I think the way you would do it would be to use one of the UART outputs of the pyboard microcontroller, which is a two wire serial interface with seperate recieve (RX) and transmit (TX) pins, controlled from micropython using the [pyb.UART](https://docs.micropython.org/en/latest/library/pyb.UART.html) class.  The output from the pyboard uses 0 and 3.3V for the low and high logic levels, but the actuators use the RS323 standard which uses ~ +/-10V logic levels.  You would therefore need to use a logic level converter IC to convert from the low voltage UART signals from the pyboard to the higher voltages expected by the actuator.  One option would be the MAX232 chip which [sparkfun](https://www.sparkfun.com/products/11189) have a breakout board for, though this appears to be out of stock currently.  

Several of the pyControl breakout board ports have UARTs on the two digital intput output lines.  You could make the physical connection to these lines using using a pyControl port adapter board (see the [docs](https://pycontrol.readthedocs.io/en/latest/user-guide/hardware/#behaviour-ports)) which has a screw terminal to make it easy to connect wires to a port.  For prototyping you could implement the circuit with the level converter IC on a breadboard.

A couple of relevent stack overflow posts about UART and RS232:

https://electronics.stackexchange.com/questions/110478/difference-between-uart-and-rs-232

https://stackoverflow.com/questions/55237103/uart-to-rs-232-convension
